![Py4Eng](img/logo.png)

# Modules: import and install
## Yoav Ram

# Modules

Luckily, we don't have to create all the code alone. Many people have written functions that perform various tasks.  
These functions are grouped into packages, called _modules_, which are usually designed to address a specific set of tasks.  
These functions are not always 'ready to use' by our code, but rather have to be _imported_ into it.

## Python version

We start with a simple example. Suppose we want to check the version of Python we are using. For this we would need to load the `sys` module using the `import` statement:

In [1]:
import sys

We can get a `version_info` object that has details on the specific source code used to compile Python:

In [2]:
v = sys.version_info
print(v)
print("We are using Python ", v.major, ".", v.minor, sep='')

sys.version_info(major=3, minor=7, micro=3, releaselevel='final', serial=0)
We are using Python 3.7


We can get more details on the specific _build_ we are using:

In [3]:
print(sys.version)

3.7.3 | packaged by conda-forge | (default, Mar 27 2019, 15:43:19) 
[Clang 4.0.1 (tags/RELEASE_401/final)]


`sys` also has other details on the system we are running on, look it up at the [docs](https://docs.python.org/3.5/library/sys.html). There is also a `platform` module:

In [10]:
import platform
print('Platform', platform.system())
print('Version', platform.version())
print('Architecture', platform.architecture())
print('Processor', platform.processor())

Platform Darwin
Version Darwin Kernel Version 18.5.0: Mon Mar 11 20:40:32 PDT 2019; root:xnu-4903.251.3~3/RELEASE_X86_64
Architecture ('64bit', '')
Processor i386


## Math module

Suppose we want to do some trigonometry, we can (in principal) write our own sine, cosine, etc. But it would be much easier to use the built-in _math_ module.

First, we import the module:

In [11]:
import math

Now we can use values and functions within the math module

In [13]:
two_pi = 2 * math.pi
print(two_pi)
radius = 2.3
perimeter = two_pi * radius
print(perimeter)
print(math.sin(math.pi / 6))
print(math.cos(math.pi / 3))

6.283185307179586
14.451326206513047
0.49999999999999994
0.5000000000000001


To see what's inside the module, we can call the `dir` and `help` functions on it, or just [search the web](https://duckduckgo.com/?q=python+math+module).

In [14]:
print(dir(math))

['__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'ceil', 'copysign', 'cos', 'cosh', 'degrees', 'e', 'erf', 'erfc', 'exp', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'modf', 'nan', 'pi', 'pow', 'radians', 'remainder', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'tau', 'trunc']


In [15]:
print(help(math))

Help on module math:

NAME
    math

MODULE REFERENCE
    https://docs.python.org/3.7/library/math
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    This module is always available.  It provides access to the
    mathematical functions defined by the C standard.

FUNCTIONS
    acos(x, /)
        Return the arc cosine (measured in radians) of x.
    
    acosh(x, /)
        Return the inverse hyperbolic cosine of x.
    
    asin(x, /)
        Return the arc sine (measured in radians) of x.
    
    asinh(x, /)
        Return the inverse hyperbolic sine of x.
    
    atan(x, /)
        Return the arc tangent (measured in radians) of x.
    
    atan2(y, x, /)
        Return the arc tangent (measured

If you only need one or two functions from a module, you can import them, instead of the whole module. This way, we don't have to call the module name every time.

In [16]:
from math import pi, sin, cos

In [18]:
# now we can use values and functions within the math module
two_pi = 2 * pi
print(two_pi)
print(sin(pi / 6))
print(cos(pi / 3))

6.283185307179586
0.49999999999999994
0.5000000000000001


We can import using an alias with the `as` command. This is especially useful if you have two module with the same API and you want to change between them.

You can also view the module documentation to see what functions are available and how to use them. Each python module has a documentation page, for example: https://docs.python.org/3/library/math.html

Two more useful links:
- [PyPI](https://pypi.python.org/pypi): a repository of (almost) all Python modules  
- [List of scientific python modules](https://wiki.python.org/moin/NumericAndScientific)

## Installing packages

Many Python modules are included within the core distribution, and all you have to do is `import` them. 
However, many more modules are contained in 3rd party packages that must be downloaded and installed first.  

Python pacakges can be installed by downloading the package source code and then running a special script that comes with the pacakge, `python setup.py install`. But this is not the recommended way to install packages.

To manage installations and dependencies, most people use a pacakge manager. The most common one is [pip](https://pip.pypa.io/) (which replaced `easy_install` some years ago), which is the package manager recommended by the Python organization. _pip_ installs packages that were uploaded (e.g. _deployed_) to [PyPI](http://pypi.python.org/), the Python Package Index, which is the main repository for Python packages.

However, scientific pacakges (such as Numpy and Scipy) sometimes have non-Python dependencies, mostly C and Fortran code, and these dependecies can be hard to compile and link - especially on standard Windows machines. Therefore, [Continuum](http://continuum.io/) started the [conda](http://conda.pydata.org/) project. *conda* is an alternative to *pip*, able to install Python packages, but it can also do more, as it can install non-Python software (even *R*). So *conda* can install Python packages with non-Python dependencies much more smoothly.

If we installed Python using the Anaconda distribution (also by *Continuum*) then we already have *pip* and *conda*.
Let's start by updating them (remember that prefixing commands with `!` in the notebook runs them as system commands, so we don't have to open a console window):

In [11]:
!conda update conda pip -y -q

Fetching package metadata ...........
Solving package specifications: ..........

# All requested packages already installed.
# packages in environment at /Users/yoavram/miniconda3:
#
conda                     4.2.13                   py35_0    conda-forge
pip                       9.0.1                    py35_0    conda-forge


### Install with *conda*

We will prefer to use *conda* as it has more *power*.
Let's install the *seaborn* package - we'll use it at a later occasion.
The `-q` option tells conda to supress progress bars, `-y` tells it to go ahead without prompting us for approval.
If you are running *conda* from a console, you can skip these options and see the extra output, but when running from inside a notebook it won't work without `-y`.

In [12]:
!conda install seaborn -y -q

Fetching package metadata ...........
Solving package specifications: ..........

Package plan for installation in environment /Users/yoavram/miniconda3:

The following NEW packages will be INSTALLED:

    cycler:          0.10.0-py35_0                      conda-forge
    matplotlib:      2.0.0-np112py35_2                  conda-forge
    numpy:           1.12.0-py35_blas_openblas_200      conda-forge [blas_openblas]
    pandas:          0.19.2-np112py35_1                 conda-forge
    patsy:           0.4.1-py35_0                       conda-forge
    pyparsing:       2.1.10-py35_0                      conda-forge
    python-dateutil: 2.6.0-py35_0                       conda-forge
    pytz:            2016.10-py35_0                     conda-forge
    scipy:           0.18.1-np112py35_blas_openblas_201 conda-forge [blas_openblas]
    seaborn:         0.7.1-py35_0                       conda-forge
    statsmodels:     0.8.0-np112py35_0                  conda-forge



*conda* installs packages from the [Anaconda.org]() repository, which is different from *PyPI* and has less packages. So some pacakges *conda* just wouldn't find and then we will try *pip*. 

### Installing with *pip*
Let's install another package we will need later on, *watchdog*:

In [13]:
!pip install watchdog

### Download pre-compiled installers

If nothing else works, you can try looking for your package in [Gohlke's website](http://www.lfd.uci.edu/~gohlke/pythonlibs/). It contains many downloadable installers(`exe` files you just click through to easily install a package) and `wheel` files which you install by calling `pip install filename.wheel`. Make sure to choose the download that fits your python version and operating system.  

### Installing from source code

Not all modules are available through this website. If you don't find your module here, you might have to install from source. Details [here](https://packaging.python.org/installing/)

# References

- [Detailed guide on installing Python modules with pip](https://packaging.python.org/installing/)
- [Packaging and Deployment with Conda](https://speakerdeck.com/teoliphant/packaging-and-deployment-with-conda) by Travis E. Oliphant

## Colophon
This notebook was written by [Yoav Ram](http://python.yoavram.com) and is part of the [_Python for Engineers_](https://github.com/yoavram/Py4Eng) course.

The notebook was written using [Python](http://python.org/) 3.7.
Dependencies listed in [environment.yml](../environment.yml), full versions in [environment_full.yml](../environment_full.yml).

This work is licensed under a CC BY-NC-SA 4.0 International License.

![Python logo](https://www.python.org/static/community_logos/python-logo.png)